lib/minstall.js

```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
+    // it is installed. Because of this we just work with the path of the first
+    // module that requests this dependency
+    const possiblePathElements = requestedDependency.requestedBy[0].substr(cwd.length).split(path.sep).filter((pathElement) => {
+      return pathElement.length > 0;
     });
-}
 
-function persistExistingConfilctingDependencies(packetsToKeep, moduleFolderName) {
+    // because we check the current path first, and then add the path element, the last element won't be checked.
+    // because of that we just add another pathElement that will be ignored, but will make the last real
+    // pathElement not be ignored
+    possiblePathElements.push('');
+    let currentPath = cwd;
+
+    for (const possiblePathElement of possiblePathElements) {
+      // is this Dependency already on the list of things to install?
+      let installModuleHere = true;
+      for (const modulePath in optimalDependencyTargetFolder) {
+        const modulesToBeInstalled = optimalDependencyTargetFolder[modulePath];
+        const matchingModule = modulesToBeInstalled.find((moduleToBeInstalled) => {
+          return moduleToBeInstalled.identifier === requestedDependency.identifier;
+        });
+
+        if (matchingModule) {
+          logger.debug(`no need to install ${requestedDependency.identifier} to ${currentPath}. a matching version will already be installed to ${modulePath}`);
+          installModuleHere = false;
+          break;
+        }
+      }
+
+      // Is a conflicting dependency-version already installed here?
+      if (installModuleHere) {
+        for (const installedDependency of alreadyInstalledDependencies) {
+          if (installedDependency.name === requestedDependency.name &&
+              installedDependency.location === path.join(currentPath, 'node_modules')) {
+            logger.debug(`${requestedDependency.identifier} can't be installed to ${currentPath}. it conflicts with the already installed ${installedDependency.name}@"${installedDependency.version}"`);
+            installModuleHere = false;
+            break;
+          }
+        }
+      }
+
+      // will any conflicting dependency-versions be installed here?
+      if (installModuleHere && optimalDependencyTargetFolder[currentPath]) {
+        const conflictingDependency = optimalDependencyTargetFolder[currentPath].find((toBeInstalledDependency) => {
+          return toBeInstalledDependency.name === requestedDependency.name &&
+                 toBeInstalledDependency.versionRange !== requestedDependency.versionRange;
+        });
+
+        if (conflictingDependency) {
+          logger.debug(`${requestedDependency.identifier} can't be installed to ${currentPath}. it'd conflict with the to be installed ${conflictingDependency.identifier}`);
+          installModuleHere = false;
+        }
+      }
+
+      if (!installModuleHere) {
+        currentPath = path.join(currentPath, possiblePathElement);
+        continue;
+      }
```

Comment:
I would strongly expect just to see `continue` here.

Why is the currentPath set here?

Reply:
and just in this case.

Reply:
the three parts above this block check if there is any reason the module cannot be installed in the location that is currently checked. If any of these checks say "nope, can't install here", we need to check the next possible location (the next "currentPath"). because of this we set the currentPath to the next one that we can check and continue